weed/mount/weedfs_file_copy_range.go

```diff
@@ -0,0 +1,120 @@
+package mount
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/hanwen/go-fuse/v2/fuse"
+
+	"github.com/seaweedfs/seaweedfs/weed/glog"
+)
+
+// CopyFileRange copies data from one file to another from and to specified offsets.
+//
+// See https://man7.org/linux/man-pages/man2/copy_file_range.2.html
+// See https://github.com/libfuse/libfuse/commit/fe4f9428fc403fa8b99051f52d84ea5bd13f3855
+/**
+ * Copy a range of data from one file to another
+ *
+ * Niels de Vos: • libfuse: add copy_file_range() support
+ *
+ * Performs an optimized copy between two file descriptors without the
+ * additional cost of transferring data through the FUSE kernel module
+ * to user space (glibc) and then back into the FUSE filesystem again.
+ *
+ * In case this method is not implemented, applications are expected to
+ * fall back to a regular file copy.   (Some glibc versions did this
+ * emulation automatically, but the emulation has been removed from all
+ * glibc release branches.)
+ */
+func (wfs *WFS) CopyFileRange(cancel <-chan struct{}, in *fuse.CopyFileRangeIn) (written uint32, code fuse.Status) {
+	// flags must equal 0 for this syscall as of now
+	if in.Flags != 0 {
+		return 0, fuse.EINVAL
+	}
+
+	// files must exist
+	fhOut := wfs.GetHandle(FileHandleId(in.FhOut))
+	if fhOut == nil {
+		return 0, fuse.EBADF
+	}
+	fhIn := wfs.GetHandle(FileHandleId(in.FhIn))
+	if fhIn == nil {
+		return 0, fuse.EBADF
+	}
+
+	// lock source and target file handles
+	fhOut.Lock()
+	defer fhOut.Unlock()
+	fhOut.entryLock.Lock()
+	defer fhOut.entryLock.Unlock()
+
+	if fhOut.entry == nil {
+		return 0, fuse.ENOENT
+	}
+
+	if fhIn.fh != fhOut.fh {
+		fhIn.Lock()
+		defer fhIn.Unlock()
+		fhIn.entryLock.Lock()
+		defer fhIn.entryLock.Unlock()
+	}
+
+	// directories are not supported
+	if fhIn.entry.IsDirectory || fhOut.entry.IsDirectory {
+		return 0, fuse.EISDIR
+	}
+
+	// cannot copy data to an overlapping range of the same file
+	offInEnd := in.OffIn + in.Len - 1
+	offOutEnd := in.OffOut + in.Len - 1
+
+	if fhIn.inode == fhOut.inode && in.OffIn <= offOutEnd && offInEnd >= in.OffOut {
```

Comment:
Seems not a hard requirement: We can read from the input first, and write the data to an overlapping range in the same file.

Reply:
ok sure, why not :)